{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/qbio/nest/alpaca/Tutorial\n"
     ]
    }
   ],
   "source": [
    "%cd /qbio/nest/alpaca/Tutorial\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "\n",
    "\n",
    "#For each gene, extract data of its longest transcript\n",
    "#This function returns a dictionary. key: gene name(GENCODE) value: [start position, end position, length of transcript, direction(+/-)]\n",
    "def makeTrxdata(gtf):\n",
    "    with open(gtf) as inpfile:\n",
    "        gt_dict={}\n",
    "        for line in inpfile:\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            feature=line.strip('\\n').split('\\t')\n",
    "            strand=feature[6]\n",
    "            #print(strand)\n",
    "            \n",
    "            \n",
    "            \n",
    "            feature=line.strip('\\n').split('\\t')\n",
    "            contig=feature[0]\n",
    "            feature_type=feature[2]\n",
    "            start=int(feature[3])\n",
    "            end=int(feature[4])\n",
    "            \n",
    "            strand=feature[6]\n",
    "\n",
    "            attribute_dict={}\n",
    "            \n",
    "            for attribute in feature[-1].split(';')[:-1]:\n",
    "                atr=attribute.strip(' ').split(' ')\n",
    "                attribute_dict[atr[0]]=atr[1].strip('\"')\n",
    "                \n",
    "            #print(contig, feature_type, start,end,strand)\n",
    "            #print(attribute_dict)\n",
    "            #break\n",
    "            if attribute_dict['gene_type'] !='protein_coding':\n",
    "                continue\n",
    "            else:\n",
    "                gene_id=attribute_dict['gene_id'].strip()\n",
    "                if feature_type=='gene':\n",
    "                    gt_dict[gene_id]={}\n",
    "                elif feature_type=='transcript':\n",
    "                    t_id=attribute_dict['transcript_id']\n",
    "                    gt_dict[gene_id][t_id]=[start, end, abs(start-end)+1, strand]\n",
    "        return(gt_dict)\n",
    "                    #print(gt_dict)\n",
    "            #        break\n",
    "            \n",
    "            #    if attribute_dict['gene_type']=='protein_coding':                \n",
    "\n",
    "\n",
    "#Similarly, extract data of each gene.\n",
    "\n",
    "def get_longest_isoform(gt_dict):\n",
    "    g_lt_dict={}\n",
    "    for g, ts in gt_dict.items():\n",
    "        argmax=np.argmax(np.array(ts.values()))\n",
    "        \n",
    "        #print(list(ts.keys())[argmax])\n",
    "        #break\n",
    "        \n",
    "#        break\n",
    "        g_lt_dict[g]=list(ts.keys())[argmax]\n",
    "        #break\n",
    "    return(g_lt_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run the functions defined above, and assign their results\n",
    "\n",
    "#Gene = makeGenedata('./GenomeIndex/gencode.vM25.primary_assembly.annotation.gtf')\n",
    "Trxdata = makeTrxdata('./GenomeIndex/gencode.vM25.primary_assembly.annotation.gtf')\n",
    "longest_isoform=get_longest_isoform(Trxdata)\n",
    "#print(Trxdata)\n",
    "#print(longest_isoform)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21936\n",
      "19974\n"
     ]
    }
   ],
   "source": [
    "#This function returns the data about exons and CDSs for each transcript.\n",
    "\n",
    "def makeExonData(addr, Trxdata):\n",
    "    result={}\n",
    "    count=0\n",
    "    with open(addr) as file_data:\n",
    "        for line in file_data:\n",
    "            exonlength=0\n",
    "            status=0\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else:\n",
    "                tmp1 = line.split('\\t')\n",
    "                if 'protein_coding' in tmp1[8] and 'exon' in tmp1[2]:\n",
    "                    tmp2 = tmp1[8].split(';')\n",
    "                    #tmp1: ['chr1', 'HAVANA', 'exon', '3213609', '3216344', '.', '-', '.', 'gene_id \"ENSMUSG00000051951.5\"; transcript_id \"ENSMUST00000162897.1\"; gene_type \"protein_coding\"; gene_name \"Xkr4\"; transcript_type \"processed_transcript\"; transcript_name \"Xkr4-203\"; exon_number 1; exon_id \"ENSMUSE00000858910.1\"; level 2; transcript_support_level \"1\"; mgi_id \"MGI:3528744\"; havana_gene \"OTTMUSG00000026353.2\"; havana_transcript \"OTTMUST00000086625.1\";\\n']\n",
    "                    #tmp2: ['gene_id \"ENSMUSG00000051951.5\"', ' transcript_id \"ENSMUST00000162897.1\"', ' gene_type \"protein_coding\"', ' gene_name \"Xkr4\"', ' transcript_type \"processed_transcript\"', ' transcript_name \"Xkr4-203\"', ' exon_number 1', ' exon_id \"ENSMUSE00000858910.1\"', ' level 2', ' transcript_support_level \"1\"', ' mgi_id \"MGI:3528744\"', ' havana_gene \"OTTMUSG00000026353.2\"', ' havana_transcript \"OTTMUST00000086625.1\"', '\\n']\n",
    "                    trx_id = tmp2[1][16:-1]\n",
    "                    gene_id = tmp2[0][9:-1]\n",
    "                    exon_id = tmp2[7][10:-1]\n",
    "        \n",
    "                    if trx_id in longest_isoform.values():\n",
    "                        left=int(tmp1[3])\n",
    "                        right=int(tmp1[4])\n",
    "                        position = [left, right, abs(left-right)+1, tmp1[6]]\n",
    "                        \n",
    "                        if gene_id not in result:\n",
    "                            result[gene_id] = {}\n",
    "                            result[gene_id][exon_id] = position                             \n",
    "                        else:\n",
    "                            result[gene_id][exon_id] = position\n",
    "                        count+=1\n",
    "                    else:\n",
    "                        continue\n",
    "            #if count>70:\n",
    "             #    break\n",
    "    \n",
    "    return result\n",
    "\n",
    "\n",
    "def makeCDSData(addr, Trxdata):\n",
    "    result={}\n",
    "    count=0\n",
    "    with open(addr) as file_data:\n",
    "        for line in file_data:\n",
    "            exonlength=0\n",
    "            status=0\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else:\n",
    "                tmp1 = line.split('\\t')\n",
    "                if 'protein_coding' in tmp1[8] and 'CDS' in tmp1[2]:\n",
    "                    tmp2 = tmp1[8].split(';')\n",
    "                    #tmp1: ['chr1', 'HAVANA', 'exon', '3213609', '3216344', '.', '-', '.', 'gene_id \"ENSMUSG00000051951.5\"; transcript_id \"ENSMUST00000162897.1\"; gene_type \"protein_coding\"; gene_name \"Xkr4\"; transcript_type \"processed_transcript\"; transcript_name \"Xkr4-203\"; exon_number 1; exon_id \"ENSMUSE00000858910.1\"; level 2; transcript_support_level \"1\"; mgi_id \"MGI:3528744\"; havana_gene \"OTTMUSG00000026353.2\"; havana_transcript \"OTTMUST00000086625.1\";\\n']\n",
    "                    #tmp2: ['gene_id \"ENSMUSG00000051951.5\"', ' transcript_id \"ENSMUST00000162897.1\"', ' gene_type \"protein_coding\"', ' gene_name \"Xkr4\"', ' transcript_type \"processed_transcript\"', ' transcript_name \"Xkr4-203\"', ' exon_number 1', ' exon_id \"ENSMUSE00000858910.1\"', ' level 2', ' transcript_support_level \"1\"', ' mgi_id \"MGI:3528744\"', ' havana_gene \"OTTMUSG00000026353.2\"', ' havana_transcript \"OTTMUST00000086625.1\"', '\\n']\n",
    "                    trx_id = tmp2[1][16:-1]\n",
    "                    gene_id = tmp2[0][9:-1]\n",
    "                    exon_id = tmp2[7][10:-1]\n",
    "        \n",
    "                    if trx_id in longest_isoform.values():\n",
    "                        left=int(tmp1[3])\n",
    "                        right=int(tmp1[4])\n",
    "                        position = [left, right, abs(left-right)+1, tmp1[6]]\n",
    "                        \n",
    "                        if gene_id not in result:\n",
    "                            result[gene_id] = {}\n",
    "                            result[gene_id][exon_id] = position                             \n",
    "                        else:\n",
    "                            result[gene_id][exon_id] = position\n",
    "                        count+=1\n",
    "                    else:\n",
    "                        continue\n",
    "            #if count>70:\n",
    "             #    break\n",
    "    \n",
    "    return result\n",
    "\n",
    "Exondata = makeExonData('./GenomeIndex/gencode.vM25.primary_assembly.annotation.gtf', Trxdata)\n",
    "print(len(Exondata))\n",
    "CDSdata = makeCDSData('./GenomeIndex/gencode.vM25.primary_assembly.annotation.gtf', Trxdata)\n",
    "print(len(CDSdata))\n",
    "#print(Length)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Siamilarly, extract data of starting/stop codon of each transcript.\n",
    "\n",
    "def makeStartData(addr, Trxdata):\n",
    "    result={}\n",
    "    count=0\n",
    "    with open(addr) as file_data:\n",
    "        for line in file_data:\n",
    "            exonlength=0\n",
    "            status=0\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else:\n",
    "                tmp1 = line.split('\\t')\n",
    "                if 'protein_coding' in tmp1[8] and 'start_codon' in tmp1[2]:\n",
    "                    tmp2 = tmp1[8].split(';')\n",
    "                    #tmp1: ['chr1', 'HAVANA', 'exon', '3213609', '3216344', '.', '-', '.', 'gene_id \"ENSMUSG00000051951.5\"; transcript_id \"ENSMUST00000162897.1\"; gene_type \"protein_coding\"; gene_name \"Xkr4\"; transcript_type \"processed_transcript\"; transcript_name \"Xkr4-203\"; exon_number 1; exon_id \"ENSMUSE00000858910.1\"; level 2; transcript_support_level \"1\"; mgi_id \"MGI:3528744\"; havana_gene \"OTTMUSG00000026353.2\"; havana_transcript \"OTTMUST00000086625.1\";\\n']\n",
    "                    #tmp2: ['gene_id \"ENSMUSG00000051951.5\"', ' transcript_id \"ENSMUST00000162897.1\"', ' gene_type \"protein_coding\"', ' gene_name \"Xkr4\"', ' transcript_type \"processed_transcript\"', ' transcript_name \"Xkr4-203\"', ' exon_number 1', ' exon_id \"ENSMUSE00000858910.1\"', ' level 2', ' transcript_support_level \"1\"', ' mgi_id \"MGI:3528744\"', ' havana_gene \"OTTMUSG00000026353.2\"', ' havana_transcript \"OTTMUST00000086625.1\"', '\\n']\n",
    "                    trx_id = tmp2[1][16:-1]\n",
    "                    gene_id = tmp2[0][9:-1]\n",
    "                    exon_id = tmp2[7][10:-1]\n",
    "        \n",
    "                    if trx_id in longest_isoform.values():\n",
    "                        left=int(tmp1[3])\n",
    "                        right=int(tmp1[4])\n",
    "                        position = [left, right, abs(left-right)+1, tmp1[6]]\n",
    "                        \n",
    "                        if gene_id not in result:\n",
    "                            result[gene_id] = {}\n",
    "                            result[gene_id][trx_id] = position                             \n",
    "                        else:\n",
    "                            result[gene_id][trx_id] = position\n",
    "                        count+=1\n",
    "                    else:\n",
    "                        continue\n",
    "            #if count>70:\n",
    "             #    break\n",
    "    \n",
    "    return result\n",
    "\n",
    "\n",
    "def makeStopData(addr, Trxdata):\n",
    "    result={}\n",
    "    count=0\n",
    "    with open(addr) as file_data:\n",
    "        for line in file_data:\n",
    "            exonlength=0\n",
    "            status=0\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else:\n",
    "                tmp1 = line.split('\\t')\n",
    "                if 'protein_coding' in tmp1[8] and 'stop_codon' in tmp1[2]:\n",
    "                    tmp2 = tmp1[8].split(';')\n",
    "                    #tmp1: ['chr1', 'HAVANA', 'exon', '3213609', '3216344', '.', '-', '.', 'gene_id \"ENSMUSG00000051951.5\"; transcript_id \"ENSMUST00000162897.1\"; gene_type \"protein_coding\"; gene_name \"Xkr4\"; transcript_type \"processed_transcript\"; transcript_name \"Xkr4-203\"; exon_number 1; exon_id \"ENSMUSE00000858910.1\"; level 2; transcript_support_level \"1\"; mgi_id \"MGI:3528744\"; havana_gene \"OTTMUSG00000026353.2\"; havana_transcript \"OTTMUST00000086625.1\";\\n']\n",
    "                    #tmp2: ['gene_id \"ENSMUSG00000051951.5\"', ' transcript_id \"ENSMUST00000162897.1\"', ' gene_type \"protein_coding\"', ' gene_name \"Xkr4\"', ' transcript_type \"processed_transcript\"', ' transcript_name \"Xkr4-203\"', ' exon_number 1', ' exon_id \"ENSMUSE00000858910.1\"', ' level 2', ' transcript_support_level \"1\"', ' mgi_id \"MGI:3528744\"', ' havana_gene \"OTTMUSG00000026353.2\"', ' havana_transcript \"OTTMUST00000086625.1\"', '\\n']\n",
    "                    trx_id = tmp2[1][16:-1]\n",
    "                    gene_id = tmp2[0][9:-1]\n",
    "                    exon_id = tmp2[7][10:-1]\n",
    "        \n",
    "                    if trx_id in longest_isoform.values():\n",
    "                        left=int(tmp1[3])\n",
    "                        right=int(tmp1[4])\n",
    "                        position = [left, right, abs(left-right)+1, tmp1[6]]\n",
    "                        \n",
    "                        if gene_id not in result:\n",
    "                            result[gene_id] = {}\n",
    "                            result[gene_id][trx_id] = position                             \n",
    "                        else:\n",
    "                            result[gene_id][trx_id] = position\n",
    "                        count+=1\n",
    "                    else:\n",
    "                        continue\n",
    "            #if count>70:\n",
    "             #    break\n",
    "    \n",
    "    return result \n",
    "\n",
    "Startdata=makeStartData('./GenomeIndex/gencode.vM25.primary_assembly.annotation.gtf', Trxdata)\n",
    "Stopdata=makeStopData('./GenomeIndex/gencode.vM25.primary_assembly.annotation.gtf', Trxdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function returns a dictionary contains position data of sequencing result\n",
    "\n",
    "def makedata(addr, Trxdata, longest_isoform):\n",
    "    count=0\n",
    "    result = {}\n",
    "    \n",
    "    with open(addr) as file_data:\n",
    "\n",
    "        for line in file_data:\n",
    "            if line.startswith('@'):\n",
    "                continue\n",
    "            else:\n",
    "                tmp1 = line.split('\\t')\n",
    "                if int(tmp1[1]) == 4 or int(tmp1[1]) == 16 or int(tmp1[1]) == 32:\n",
    "                    continue\n",
    "                else:\n",
    "                    tmp2 = tmp1[2].split('|')\n",
    "                    #tmp2: ['ENSMUST00000198477.2', 'ENSMUSG00000106106.3', 'OTTMUSG00000053277.2', 'OTTMUST00000133068.2', 'CT010467.1-201', 'CT010467.1', '1849', 'rRNA', '']\n",
    "                    mapped_position = int(tmp1[3])\n",
    "                    if 'protein_coding' in tmp1[2]:\n",
    "                        if tmp2[0] in longest_isoform.values():\n",
    "                            count+=1\n",
    "                            if tmp2[1] in result:\n",
    "                                result[tmp2[1]].append(mapped_position)\n",
    "                            else:\n",
    "                                result[tmp2[1]]=[mapped_position]\n",
    "                        else:\n",
    "                            continue\n",
    "                    \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13820\n"
     ]
    }
   ],
   "source": [
    "#Run makedata function defined above\n",
    "\n",
    "pos_siLin = makedata('./Align_RiboProf-siLin28a/Aligned.out.sam', Trxdata, longest_isoform)\n",
    "print(len(pos_siLin))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13939\n"
     ]
    }
   ],
   "source": [
    "pos_siLuc = makedata('./Align_RiboProf-siLuc/Aligned.out.sam', Trxdata, longest_isoform)\n",
    "print(len(pos_siLuc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function calculates Relative position to start codon of 5'-end of reads, thus, x-coordinate of figure S5A, with single gene.\n",
    "#The output format is a 3D list. each element is a list with 4 elements, which are lists of relative positions.\n",
    "\n",
    "def getExon(Exons, value):\n",
    "    for exon in Exons:\n",
    "        if value > Exons[exon][0] and value < Exons[exon][1]:\n",
    "            return exon\n",
    "    else:\n",
    "        return {}\n",
    "\n",
    "#print(getExon(Exondata['ENSMUSG00000033845.13'], 4785575))\n",
    "        \n",
    "def sampledata(sample, Trxdata, longest_isoform, Exondata, CDSData, StartData, StopData, pos_siLin, pos_siLuc):\n",
    "    resdata=[]\n",
    "    for i in sample:\n",
    "        \n",
    "        Trx_name = longest_isoform[i]\n",
    "\n",
    "        Transcript = Trxdata[i][longest_isoform[i]]\n",
    "        #Transcript: [3205901, 3216344, 10443, '-']\n",
    "        Exon = Exondata[i]\n",
    "        CDS = CDSdata[i]\n",
    "        #Exon/CDS: {'ENSMUSE00000450421.7': [4785573, 4785710, 138, '-'], 'ENSMUSE00001229301.1': [4783951, 4784105, 155, '-'], 'ENSMUSE00001268906.1': [4782568, 4782733, 166, '-'], 'ENSMUSE00001298601.1': [4777525, 4777648, 124, '-'], 'ENSMUSE00000771965.1': [4773206, 4774516, 1311, '-']}\n",
    "        \n",
    "        Exonlength = 0\n",
    "        for exon in Exon:\n",
    "            Exonlength += Exon[exon][2]\n",
    "            \n",
    "        #print(Exonlength)\n",
    "\n",
    "        \n",
    "        Start = StartData[i]\n",
    "        Stop = StopData[i]\n",
    "        siLin_position = pos_siLin[i]\n",
    "        siLuc_position = pos_siLuc[i]\n",
    "        \n",
    "        #exonlength = 0\n",
    "        #for exon in Exon:\n",
    "        #    exonlength += Exon[exon][2]\n",
    "        \n",
    "        Linpos_start = []\n",
    "        Linpos_stop = []\n",
    "        Lucpos_start = []\n",
    "        Lucpos_stop = []\n",
    "        \n",
    "        start_position = Start[Trx_name][0]\n",
    "        stop_position = Stop[Trx_name][0] \n",
    "        start_exon = getExon(Exon, start_position)\n",
    "        stop_exon = getExon(Exon, stop_position)\n",
    "        \n",
    "        if len(start_exon) == 0 or len(stop_exon)==0:\n",
    "            continue\n",
    "        \n",
    "        if i in CDSdata:\n",
    "        \n",
    "            if Transcript[3] == '+':\n",
    "            \n",
    "                UTR5 = CDS[start_exon][0] - Exon[start_exon][0]\n",
    "                UTR3 = Exon[stop_exon][1] - CDS[stop_exon][1]\n",
    "                \n",
    "                for pos in siLin_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Linpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Linpos_stop.append(distance_from_stop)\n",
    "                        \n",
    "                for pos in siLuc_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Lucpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Lucpos_stop.append(distance_from_stop)                    \n",
    "           \n",
    "            else:\n",
    "                \n",
    "                UTR5 = Exon[start_exon][1] - CDS[start_exon][1]\n",
    "                UTR3 = CDS[stop_exon][0] - Exon[stop_exon][0]\n",
    "                \n",
    "                for pos in siLin_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Linpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Linpos_stop.append(distance_from_stop)\n",
    "                        \n",
    "                for pos in siLuc_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Lucpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Lucpos_stop.append(distance_from_stop)\n",
    "            \n",
    "            Linpos_start.sort()\n",
    "            Linpos_stop.sort()\n",
    "            Lucpos_start.sort()\n",
    "            Lucpos_stop.sort()\n",
    "            \n",
    "            resdata.append([Linpos_start, Linpos_stop, Lucpos_start, Lucpos_stop])\n",
    "\n",
    "    return resdata\n",
    "\n",
    "sample = ['ENSMUSG00000033845.13', 'ENSMUSG00000026980.15', 'ENSMUSG00000037169.15', 'ENSMUSG00000060377.4', 'ENSMUSG00000079435.9']\n",
    "resdata = sampledata(sample, Trxdata, longest_isoform, Exondata, CDSdata, Startdata, Stopdata, pos_siLin, pos_siLuc)\n",
    "#for i in range(4):\n",
    "    #print(test[0][i])\n",
    "#    print()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-194-50675e954bae>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m221\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mylabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'\\nsiLuc\\nRaw read count\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgrid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbAAAACGCAYAAAC8Gt8uAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAALRUlEQVR4nO3dXYhc533H8e+vkgWNa+IkXrtBL41a1LouxOBMFbdJG7vFqWQaRMAXckMMJiDcxqX0osT0wrnoTUtuSlonQhgRchHrorETFfwKpVWo41ar4hfJicJWSaNFAVu2cahTKuT8ezFHMF3vas/Mrnb3yXw/MGjO8zL7nwft/PacmX02VYUkSa35ufUuQJKkSRhgkqQmGWCSpCYZYJKkJhlgkqQmGWCSpCYtG2BJDid5JcnJJfqT5ItJ5pK8mOSWkb49SU53fQ+sZuGSpOnW5wzsK8Cey/TvBXZ1twPAlwGSbAIe6vpvAu5OctNKipUk6ZJlA6yqjgGvX2bIPuCrNfQccG2S9wO7gbmqOlNVF4Aj3VhJklZsNd4D2wqcHTme79qWapckacU2r8JjZJG2ukz74g+SHGB4CZKrr776QzfeeOMqlCZJ2shOnDhxvqpmJpm7GgE2D2wfOd4GnAO2LNG+qKo6BBwCGAwGNTs7uwqlSZI2siT/Nenc1biEeBS4p/s04q3Am1X1I+A4sCvJziRbgP3dWEmSVmzZM7AkjwC3AdclmQc+D1wFUFUHgceBO4E54CfAvV3fxST3A08Bm4DDVXXqCjwHSdIUWjbAquruZfoL+OwSfY8zDDhJklaVO3FIkppkgEmSmmSASZKaZIBJkppkgEmSmmSASZKaZIBJkppkgEmSmmSASZKaZIBJkppkgEmSmmSASZKaZIBJkppkgEmSmmSASZKaZIBJkprUK8CS7ElyOslckgcW6f+LJM93t5NJ3k7y3q7vB0le6vpmV/sJSJKm07J/kTnJJuAh4A5gHjie5GhVvXxpTFV9AfhCN/4TwJ9X1esjD3N7VZ1f1colSVOtzxnYbmCuqs5U1QXgCLDvMuPvBh5ZjeIkSVpKnwDbCpwdOZ7v2t4hybuAPcDXR5oLeDrJiSQHJi1UkqRRy15CBLJIWy0x9hPAvy64fPiRqjqX5HrgmSTfrapj7/giw3A7ALBjx44eZUmSplmfM7B5YPvI8Tbg3BJj97Pg8mFVnev+fQV4jOElyXeoqkNVNaiqwczMTI+yJEnTrE+AHQd2JdmZZAvDkDq6cFCSdwMfA7450nZ1kmsu3Qc+DpxcjcIlSdNt2UuIVXUxyf3AU8Am4HBVnUpyX9d/sBv6SeDpqnprZPoNwGNJLn2tr1XVk6v5BCRJ0ylVS72dtX4Gg0HNzvorY5L0sy7JiaoaTDLXnTgkSU0ywCRJTTLAJElNMsAkSU0ywCRJTTLAJElNMsAkSU0ywCRJTTLAJElNMsAkSU0ywCRJTTLAJElNMsAkSU0ywCRJTTLAJElNMsAkSU3qFWBJ9iQ5nWQuyQOL9N+W5M0kz3e3B/vOlSRpEpuXG5BkE/AQcAcwDxxPcrSqXl4w9FtV9YcTzpUkaSx9zsB2A3NVdaaqLgBHgH09H38lcyVJWlKfANsKnB05nu/aFvqtJC8keSLJb4w5lyQHkswmmX311Vd7lCVJmmZ9AiyLtNWC4/8Afqmqbgb+DvjGGHOHjVWHqmpQVYOZmZkeZUmSplmfAJsHto8cbwPOjQ6oqh9X1X939x8HrkpyXZ+5kiRNok+AHQd2JdmZZAuwHzg6OiDJLyZJd39397iv9ZkrSdIklv0UYlVdTHI/8BSwCThcVaeS3Nf1HwTuAv44yUXgf4D9VVXAonOv0HORJE2RDHNmYxkMBjU7O7veZUiSrrAkJ6pqMMlcd+KQJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1qVeAJdmT5HSSuSQPLNL/qSQvdrdnk9w80veDJC8leT6Jf+RLkrQqlv2LzEk2AQ8BdwDzwPEkR6vq5ZFh3wc+VlVvJNkLHAI+PNJ/e1WdX8W6JUlTrs8Z2G5grqrOVNUF4Aiwb3RAVT1bVW90h88B21a3TEmS/r8+AbYVODtyPN+1LeUzwBMjxwU8neREkgPjlyhJ0jstewkRyCJttejA5HaGAfbRkeaPVNW5JNcDzyT5blUdW2TuAeAAwI4dO3qUJUmaZn3OwOaB7SPH24BzCwcl+SDwMLCvql671F5V57p/XwEeY3hJ8h2q6lBVDapqMDMz0/8ZSJKmUp8AOw7sSrIzyRZgP3B0dECSHcCjwKer6nsj7VcnuebSfeDjwMnVKl6SNL2WvYRYVReT3A88BWwCDlfVqST3df0HgQeB9wFfSgJwsaoGwA3AY13bZuBrVfXkFXkmkqSpkqpF385aV4PBoGZn/ZUxSfpZl+REd8IzNnfikCQ1yQCTJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1yQCTJDXJAJMkNckAkyQ1qVeAJdmT5HSSuSQPLNKfJF/s+l9MckvfuZIkTWLZAEuyCXgI2AvcBNyd5KYFw/YCu7rbAeDLY8yVJGlsfc7AdgNzVXWmqi4AR4B9C8bsA75aQ88B1yZ5f8+5kiSNrU+AbQXOjhzPd219xvSZK0nS2Db3GJNF2qrnmD5zhw+QHGB4+RHgf5Oc7FGbhq4Dzq93EQ1xvcbjeo3H9RrPr006sU+AzQPbR463Aed6jtnSYy4AVXUIOASQZLaqBj1qE67XuFyv8bhe43G9xpNkdtK5fS4hHgd2JdmZZAuwHzi6YMxR4J7u04i3Am9W1Y96zpUkaWzLnoFV1cUk9wNPAZuAw1V1Ksl9Xf9B4HHgTmAO+Alw7+XmXpFnIkmaKn0uIVJVjzMMqdG2gyP3C/hs37k9HBpz/LRzvcbjeo3H9RqP6zWeidcrw+yRJKktbiUlSWrSugXYSranmkY91utT3Tq9mOTZJDevR50bRd8tzJL8ZpK3k9y1lvVtNH3WK8ltSZ5PcirJv6x1jRtNj+/Jdyf5xyQvdGt273rUuREkOZzklaV+PWri1/uqWvMbww90/Cfwyww/av8CcNOCMXcCTzD8XbJbgX9bj1o3wq3nev028J7u/l7X6/LrNTLunxi+R3vXete9kdcLuBZ4GdjRHV+/3nU3sGZ/CfxNd38GeB3Yst61r9N6/S5wC3Byif6JXu/X6wxsJdtTTaNl16uqnq2qN7rD5xj+zt206ruF2Z8CXwdeWcviNqA+6/VHwKNV9UOAqnLNll+zAq5JEuAXGAbYxbUtc2OoqmMMn/9SJnq9X68AW8n2VNNo3LX4DMOfZqbVsuuVZCvwSeAg6vP/61eB9yT55yQnktyzZtVtTH3W7O+BX2e4ecNLwJ9V1U/XprzmTPR63+tj9FfASranmkbjbMl1O8MA++gVrWhj67Nefwt8rqreHv6APNX6rNdm4EPA7wM/D3w7yXNV9b0rXdwG1WfN/gB4Hvg94FeAZ5J8q6p+fIVra9FEr/frFWAr2Z5qGvVaiyQfBB4G9lbVa2tU20bUZ70GwJEuvK4D7kxysaq+sSYVbix9vx/PV9VbwFtJjgE3A9MaYH3W7F7gr2v4Js9cku8DNwL/vjYlNmWi1/v1uoS4ku2pptGy65VkB/Ao8Okp/qn4kmXXq6p2VtUHquoDwD8AfzKl4QX9vh+/CfxOks1J3gV8GPjOGte5kfRZsx8yPGMlyQ0MN609s6ZVtmOi1/t1OQOrFWxPNY16rteDwPuAL3VnFRdrSjcU7ble6vRZr6r6TpIngReBnwIPV9XU/sWInv/H/gr4SpKXGF4i+1xVTeUu9UkeAW4DrksyD3weuApW9nrvThySpCa5E4ckqUkGmCSpSQaYJKlJBpgkqUkGmCSpSQaYJKlJBpgkqUkGmCSpSf8H5Bdgu5QEvwkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1080x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Draw Fig S5A with single gene data.\n",
    "#sample provides some protein-coding genes that ribosome footprinting result data are aligned to.\n",
    "#By setting the value of index(0~19), we can see the graph for each gene, for test.\n",
    "index=0\n",
    "#sample=['ENSMUSG00000020717.19', 'ENSMUSG00000026980.15', 'ENSMUSG00000078812.10', 'ENSMUSG00000037169.15', 'ENSMUSG00000060377.4', \n",
    "#        'ENSMUSG00000079435.9', 'ENSMUSG00000046591.10', 'ENSMUSG00000038412.8', 'ENSMUSG00000070047.14', 'ENSMUSG00000079445.3', \n",
    "#        'ENSMUSG00000001552.14', 'ENSMUSG00000021025.8', 'ENSMUSG00000028792.14', 'ENSMUSG00000027248.13', 'ENSMUSG00000025351.14', \n",
    "#        'ENSMUSG00000030062.7', 'ENSMUSG00000063316.13', 'ENSMUSG00000060461.5', 'ENSMUSG00000041959.14', 'ENSMUSG00000025967.16']\n",
    "#resdata=sampledata(sample, Trxdata, longest_isoform, Exondata, CDSdata, Startdata, Stopdata, pos_siLin, pos_siLuc)\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (15,4)\n",
    "\n",
    "plt.subplot(221)\n",
    "plt.hist(resdata[index][2], bins=range(-50,50))\n",
    "plt.ylabel(sample[index]+'\\nsiLuc\\nRaw read count\\n')\n",
    "plt.grid()\n",
    "\n",
    "plt.subplot(222)\n",
    "plt.hist(resdata[index][3], bins=range(-50,20))\n",
    "plt.ylabel('siLin28a\\nRaw read count')\n",
    "plt.grid()\n",
    "\n",
    "plt.subplot(223)\n",
    "plt.hist(resdata[index][0], bins=range(-50,50))\n",
    "plt.ylabel('siLin28a\\nRaw read count')\n",
    "plt.grid()\n",
    "\n",
    "plt.subplot(224)\n",
    "plt.hist(resdata[index][1], bins=range(-50,20))\n",
    "plt.ylabel('siLin28a\\nRaw read count')\n",
    "plt.grid()\n",
    "\n",
    "\n",
    "#gap1 120\n",
    "#gap2 17343\n",
    "#print(Trx['ENSMUSG00000028792.14'])\n",
    "#print(Start['ENSMUSG00000028792.14'])\n",
    "#print(Stop['ENSMUSG00000028792.14'])\n",
    "#print(Lin['ENSMUSG00000028792.14'])\n",
    "#plt.savefig('Figure S5A_singlegene2.png', dpi=500)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function calculates Relative position to start codon of 5'-end of reads, thus, x-coordinate of figure S5A with whole gene data.\n",
    "#The output format is a 3D list. each element is a list with 4 elements, which are lists of relative positions.\n",
    "\n",
    "def getExon(Exons, value):\n",
    "    for exon in Exons:\n",
    "        if value > Exons[exon][0] and value < Exons[exon][1]:\n",
    "            return exon\n",
    "    else:\n",
    "        return {}\n",
    "\n",
    "#print(getExon(Exondata['ENSMUSG00000033845.13'], 4785575))\n",
    "        \n",
    "def wholedata(sample, Trxdata, longest_isoform, Exondata, CDSData, StartData, StopData, pos_siLin, pos_siLuc):\n",
    "\n",
    "    Linpos_start = []\n",
    "    Linpos_stop = []\n",
    "    Lucpos_start = []\n",
    "    Lucpos_stop = []\n",
    "\n",
    "    for i in sample:\n",
    "        if i in Trxdata and i in longest_isoform and i in Exondata and i in CDSData and i in StartData and i in StopData and i in pos_siLin and i in pos_siLuc:\n",
    "            \n",
    "            Trx_name = longest_isoform[i]\n",
    "    \n",
    "            Transcript = Trxdata[i][longest_isoform[i]]\n",
    "            #Transcript: [3205901, 3216344, 10443, '-']\n",
    "            Exon = Exondata[i]\n",
    "            CDS = CDSdata[i]\n",
    "            #Exon/CDS: {'ENSMUSE00000450421.7': [4785573, 4785710, 138, '-'], 'ENSMUSE00001229301.1': [4783951, 4784105, 155, '-'], 'ENSMUSE00001268906.1': [4782568, 4782733, 166, '-'], 'ENSMUSE00001298601.1': [4777525, 4777648, 124, '-'], 'ENSMUSE00000771965.1': [4773206, 4774516, 1311, '-']}\n",
    "            \n",
    "            Exonlength = 0\n",
    "            for exon in Exon:\n",
    "                Exonlength += Exon[exon][2]\n",
    "                \n",
    "            #print(Exonlength)\n",
    "    \n",
    "            \n",
    "            Start = StartData[i]\n",
    "            Stop = StopData[i]\n",
    "            siLin_position = pos_siLin[i]\n",
    "            siLuc_position = pos_siLuc[i]\n",
    "            \n",
    "            #exonlength = 0\n",
    "            #for exon in Exon:\n",
    "            #    exonlength += Exon[exon][2]\n",
    "            \n",
    "            start_position = Start[Trx_name][0]\n",
    "            stop_position = Stop[Trx_name][0] \n",
    "            start_exon = getExon(Exon, start_position)\n",
    "            stop_exon = getExon(Exon, stop_position)\n",
    "            \n",
    "            if len(start_exon) == 0 or len(stop_exon)==0 or start_exon not in CDS or start_exon not in Exon or stop_exon not in CDS or stop_exon not in CDS:\n",
    "                continue\n",
    "        \n",
    "            \n",
    "            if Transcript[3] == '+':  \n",
    "                \n",
    "                UTR5 = CDS[start_exon][0] - Exon[start_exon][0]\n",
    "                UTR3 = Exon[stop_exon][1] - CDS[stop_exon][1] \n",
    "                \n",
    "                for pos in siLin_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Linpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Linpos_stop.append(distance_from_stop)\n",
    "                            \n",
    "                for pos in siLuc_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Lucpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Lucpos_stop.append(distance_from_stop)                    \n",
    "               \n",
    "            else:\n",
    "                \n",
    "                UTR5 = Exon[start_exon][1] - CDS[start_exon][1]\n",
    "                UTR3 = CDS[stop_exon][0] - Exon[stop_exon][0]  \n",
    "                \n",
    "                for pos in siLin_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Linpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Linpos_stop.append(distance_from_stop)\n",
    "                            \n",
    "                for pos in siLuc_position:\n",
    "                    distance_from_start = pos - UTR5\n",
    "                    Lucpos_start.append(distance_from_start)\n",
    "                    distance_from_stop = UTR3 - (Exonlength - pos)\n",
    "                    Lucpos_stop.append(distance_from_stop)\n",
    "                \n",
    "            #Linpos_start.sort()\n",
    "            #Linpos_stop.sort()\n",
    "            #Lucpos_start.sort()\n",
    "            #Lucpos_stop.sort()\n",
    "            \n",
    "            \n",
    "\n",
    "    return [Linpos_start, Linpos_stop, Lucpos_start, Lucpos_stop]\n",
    "\n",
    "sample = ['ENSMUSG00000033845.13', 'ENSMUSG00000026980.15', 'ENSMUSG00000037169.15', 'ENSMUSG00000060377.4', 'ENSMUSG00000079435.9']\n",
    "resdata = sampledata(sample, Trxdata, longest_isoform, Exondata, CDSdata, Startdata, Stopdata, pos_siLin, pos_siLuc)\n",
    "#for i in range(4):\n",
    "    #print(test[0][i])\n",
    "#    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSMUSG00000061518.11', 'ENSMUSG00000079941.1', 'ENSMUSG00000059070.16', 'ENSMUSG00000004748.5', 'ENSMUSG00000037470.14', 'ENSMUSG00000023944.14', 'ENSMUSG00000025534.17', 'ENSMUSG00000021670.14', 'ENSMUSG00000047881.14', 'ENSMUSG00000022982.10', 'ENSMUSG00000025008.15', 'ENSMUSG00000056201.8', 'ENSMUSG00000003429.11', 'ENSMUSG00000039682.12', 'ENSMUSG00000021144.14']\n",
      "<class 'dict'>\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA6UAAAD4CAYAAAAU2UDyAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAw2ElEQVR4nO3de7RkZX3n//dHiIhIewnCKJA0xtYMihpp0VlmkkajYsxwcXnBZRS8hMSBEddgFDQ/YQ3jiCaGTExCgspFY0TGaytBRbSjWSMKKAa5ODAK2tDKKAQBB6Tx+/uj9hmqT9fpPuf0rrN3Vb1fa511qp7al+/znDr13U/tZz87VYUkSZIkSV14QNcBSJIkSZJml51SSZIkSVJn7JRKkiRJkjpjp1SSJEmS1Bk7pZIkSZKkzuzcdQCzYo899qjVq1d3HUZr7rrrLnbbbbeuw+gd22U022Vrtslo09Aul19++Y+r6pFdxzEp2siP0/C+mTMtdbEe/WI9+mWW67FQjrRTukJWr17NZZdd1nUYrdmwYQPr1q3rOozesV1Gs122ZpuMNg3tkuTGrmOYJG3kx2l438yZlrpYj36xHv0yy/VYKEc6fFeSJEmS1Bk7pZIkSZKkztgplSRJkiR1xk6pJEmSJKkzdkolSZIkSZ2xUypJkiRJ6oydUkmSJElSZ+yUSpIkSZI6Y6dUkiRJktQZO6WSJEmSpM603ilNcvxiyiS1a/WJF2zxI2mymU8lSbNiHGdKjxpRdvQY9iNJ0jQzn0qSZkJrndIkL0vyaWC/JOuHfr4E/KSt/Qzt76wktyT59lDZI5JclOS65vfDh147Kcn1Sb6T5HlD5QcmubJ57S+TpCnfJclHmvKvJVk9tM5RzT6uSzLqoEGSpGXZ0XxqfpQkTZo2z5T+T+DdwLXN77mfE4BDWtzPnHNGbPdE4OKqWgNc3Dwnyf7AkcATmnX+JslOzTpnAMcAa5qfuW2+Britqh4LnA68s9nWI4CTgacDBwEnDyd3SZJ20I7m03NGLGd+lCT1Vmud0qq6sao2VNW/q6p/Gvr5RlVtbms/Q/v7MnDrvOLDgHObx+cChw+Vn1dV91TV94DrgYOSPApYVVVfraoCPjBvnbltfRR4dvMt8fOAi6rq1qq6DbiI8XS6JUkzaEfzqflRkjRpdm57g0leyOBb0z2BND9VVava3tcIe1XVJgY73JRkz6Z8b+CSoeU2NmX3No/nl8+t84NmW5uT3A788nD5iHW2kOQYBt8ys9dee7Fhw4ZlV6xv7rzzzqmqT1u6bJcTDtjyWLVPfx/fL1uzTUazXe7Xcj6d6vw4Te+baamL9egX69Ev1mNrrXdKgXcB/6GqrhnDtpcrI8pqG+XLXWfLwqozgTMB1q5dW+vWrdtuoJNiw4YNTFN92tJluxw9b8bdG17eTRyj+H7Zmm0ymu2yhZXIp1ORH6fpfTMtdbEe/WI9+sV6bG0cs+/+qMMO6Y+aIUc0v29pyjcC+w4ttw9wc1O+z4jyLdZJsjPwUAbDoRbaliRJbWozn5ofJUm9NY5O6WXNrHwvS/LCuZ8x7GeU9dw/hf5RwKeGyo9sZgzcj8GEDV9vhjLdkeQZzfUwr5y3zty2XgR8sbmu5nPAc5M8vJnA4blNmSRJbWozn5ofJUm9NY7hu6uAnzFIRnMK+HibO0nyYWAdsEeSjQxm/DsNOD/Ja4DvAy8GqKqrkpwPXA1sBo6tqvuaTb2OwUyFuwIXNj8A7wc+mOR6Bt8AH9ls69YkpwKXNsv9l6qaP6GEJEk7aln51PwoSZo0rXdKq+pVbW9zgf28bIGXnr3A8m8H3j6i/DLgiSPK76ZJ2iNeOws4a9HBSvp/Vs+/9vW0F3QUidRvy82n5kdJ0qQZx+y7ZzNiYoOqenXb+5IkaVqZTyVJs2Icw3c/M/T4QcARONGBJElLZT6VJM2EcQzf/djw8+bali+0vR9JkqaZ+VSSNCvGMfvufGuAX1mB/UiSNM3Mp5KkqTSOa0rv4P6baBfwQ+DNbe9HkqRpZj6VJM2KcQzf3b3tbUqSNGvMp5KkWTGOiY5IcijwW83TDVX1mW0tL0mStmY+lSTNgtavKU1yGnA8gxtxXw0cn+Qdbe9HkqRpZj6VJM2KcZwp/V3gKVX1C4Ak5wLfBE4aw74kSZpW5lNJ0kwY1+y7Dxt6/NAx7UOSpImSZE2Sjya5Osl35362scrDhh6bTyVJU2kcZ0rfAXwzyZcYzBj4W/itriRJAGcDJwOnAwcDr2KQK0cxn0qSZsI4Zt/9cJINwNMYJNE3V9UP296PJEkTaNequjhJqupG4JQkX2HQUd2C+VSSNCvGMdHREcDPqmp9VX0KuDvJ4W3vR5KkCXR3kgcA1yU5rsmZe45a0HwqSZoV47im9OSqun3uSVX9KyO+AZYkaQa9AXgw8HrgQOAVwFELLGs+lSTNhHFcUzqqozuW+6FKkjRJqurS5uGdDK4n3RbzqSRpJozjTOllSf48ya8leUyS04HLx7AfSZImQpMPz0ryX5M8JMl7k3w7yf9IsnqB1cynkqSZMI5O6X8Cfg58BDgf+L/AsWPYjyRJk+Ic4FIGZ0gvAa4Fng98FjhrgXXMp5KkmbCoYUBJ/hvwruZ6FpI8HDihqv5k/rJVdRdwYptBSpI04XavqjMAkvzHqnp3U/7+JMeNWsF8KkmaFYs9U/r8uQ4pQFXdBvzuWCKSJGn6/CLJ45I8DXhwkrUASR4L7NRtaJIkdWuxEybslGSXqroHIMmuwC7jC0uSpKnyJuDTwC+Aw4GTkjwZWAX8QYdxSZLUucV2Sv8euDjJ2UABrwbOHVtUkiRNkaq6GHj8UNE/J9kDuK2q7usoLEmSemFRndKqeleSK4FnAwFOrarPDS+T5D0MOqwLbeP1OxKoJEmTLMkq4JFV9b8BqurHTfmTqupfhpYzn0qSZsqiZ9+tqgur6o1VdcL8DmnjMgZT1T8IeCpwXfPzFMBvgSVJMyvJSxjMuPuxJFc115bOOWfe4uZTSdJMWezsu3dw/7e2DwR+CbirqlbNLVNV5zbLHg0cXFX3Ns//Fvh8izFLkjRp3gIcWFWbkhwEfDDJW6rq4wxGIP0/5lNJ0qxZ7PDd3YefJzkcOGiBxR8N7A7c2jx/SFMmSdKs2qmqNgFU1deTHAx8Jsk+LDxU13wqSZoJi53oaAtV9ckkC9077TTgm0m+1Dz/beCU5exHkqQpcUeSXxu6nnRTknXAJ4EnLLCO+VSSNBMWO3z3hUNPHwCsZYFvdqvq7CQXAk9vik6sqh/uUJSSJE2217H1MN07khwCvGTUCuZTSdKsWOyZ0v8w9HgzcANw6DaWvwfYxGCShscleVxVfXlZEUqSNOGq6lsLlN8LfGgbq5pPJUlTb7HXlL5qflmSPwPeOKL8tcDxwD7AFcAzgK8Cz9qRQCVtbfWJF3QdgqQlaEYevRPYk8GZ0wA1PHHg0LLmU0nSTFj0LWFGGDnciEECfRpwY1UdDPwG8H92YD+SJE2LdwGHVtVDq2pVVe0+qkPaMJ9KkmbCjnRKs0D53VV1N0CSXarqWuDxO7AfSZKmxY+q6ppFLms+lSTNhG0O303yiIVeYuFO6cYkD2Mwo+BFSW4Dbl5ugJIkTZHLknyEQY68Z66wuV/pfOZTSdJM2N41pZczmGV3VAf03lErVNURzcNTmmnsHwp8dtkRLkOSG4A7gPuAzVW1tulgfwRYzWCippdU1W3N8icBr2mWf31Vfa4pPxA4B9gV+Efg+KqqJLsAHwAOBH4CvLSqblih6kmSJtcq4GfAc4fKCtiqUzqOfGp+lCT10TY7pVW133I2muQ3gTXNdPaPBPYGvrecbe2Ag6vqx0PPTwQurqrTmnusngi8Ocn+wJEM7hP3aOALzeyG9wFnAMcAlzBIuocAFzJI0LdV1WOTHMlg0oqXrlTFJEmTadTEgdsypnxqfpQk9cr2hu/+elVdm+Spo16vqm+MWOdkBvcxfTxwNvBLwN8Dz9zxcHfIYcC65vG5wAbgzU35eVV1D/C9JNcDBzXfJq+qqq8CJPkAcDiDpHsY99/A/KPAXyVJVY28d6skabYleVNVvSvJexhxn++qev2IdVYqn5ofJUmd2t7w3f/M4JvQdw+VDSeWUdPSH8FghsBvAFTVzUl235Egl6GAzycp4O+q6kxgr6ra1MS0KcmezbJ7M/imd87Gpuze5vH88rl1ftBsa3OS24FfBoa/eSbJMQzaj7322osNGza0VsGu3XnnnVNVn7asdLuccMDmBV/r099nuF3mx9ynOFeS/0OjTXG7zE1udNmI1xbqsI0jn05lfpym98201MV69Iv16BfrsbXtDd89pnl4BvDZqvppkv8PeCpw6gKr/by5rqQAkuzWSqRL88wmee/JYHKIa7ex7KjrZRe6jnbuwGFbr91fMEj2ZwKsXbu21q1bt82gJ8mGDRuYpvq0ZaXb5eht3Kf0hpfveBzz74N6w2kvWNZ2httlfsxtxDmJ/B8abVrbpao+3fw+d/5rzX2/RxlHPp3K/DhN75tpqYv16Bfr0S/WY2uLvSXMnzQd0t8EnsNgcoMzFlj2/CR/BzwsyR8AXwDeu8ORLkFV3dz8vgX4BHAQ8KMkjwJoft/SLL4R2Hdo9X0YzG64sXk8v3yLdZLszGDyiVvHURdJ0tRb6L7fredT86MkqY8W2ym9r/n9AuBvq+pTwAPnL5QkDGbw+yjwMQbXwbytqt7TQqyLkmS3ueFNzbfKzwW+DawHjmoWOwr4VPN4PXBkkl2S7AesAb7eDGW6I8kzmnq9ct46c9t6EfBFr5eRJC3TVmcXx5FPzY+SpL7a3jWlc25qvq39HeCdzZTvW3Vom2FGn6yqA4GLWoxzKfYCPjHIk+wM/ENVfTbJpQy+dX4N8H3gxU3MVyU5H7ga2Awc28wsCPA67p/y/sLmB+D9wAebSR9uZTA7oTTThof6LneYrzStlnrf7zHlU/OjJKmXFtspfQmD6d7/rKr+tRne88cLLHtJkqdV1aWtRLhEVfVd4Mkjyn8CPHuBdd4OvH1E+WXAE0eU302TtKVZNf96U0nbtOT7ftNyPjU/SpL6alGd0qr6GUM39m6G7mxaYPGDgT9MciNwF4MEXFX1pB2MVZKkibTM+36bTyVJM2GxZ0qX4vlj2KYkSRNrOff9xnwqSZoRrXdKq+rGtrcpqV1e/ymtuCXf99t8KkmaFYudfVeSJC3TvPt+H1ZVBwNfAm4H3thZYJIk9YCdUkmSVs5S7vstSdJMaL1TmuTVSda0vV1JkqbAou77DeZTSdLsGMdER6uB30/yqwymwP8K8JWqumIM+5IkaZIs6r7fjdWYTyVJM6D1M6VV9baqehaD+5f9M4P7mV7e9n4kSZpALwE+BxxSVf8KPIIF7vttPpUkzYrWz5Qm+RPgmcBDgG8ymMDhK23vR5KkSbOU+36bTyVJs2Icw3dfCGwGLgD+Cbikqu4ew34kjcHw7WK0ON5iR2NiPpUkzYRxDN99KvBs4OsMZha8Msk/t70fSZKmmflUkjQrxjF894nAvwd+G1gL/ACHG0kz7cqbbudoz8BKS2I+lSTNinEM330n8GXgL4FLq+reMexDkqRpZz6VJM2E1julVeUFVZJW1PzrYL2uU9PAfCpJmhXjGL67BngHsD/woLnyqnpM2/uSJGlamU8lSbNiHMN3zwZOBk4HDgZeBWQM+5G0SM6oK00k86kkaSaMo1O6a1VdnCRVdSNwSpKvMEiskrQgh+FKWzCfSpJmwjg6pXcneQBwXZLjgJuAPcewH0mSppn5VJI0E8bRKX0D8GDg9cCpwLOAV45hP5IkTbM3YD7VAkZdluHoEkmTahyz717aPLwTeFWSnYGXAl9re1+SJE0r86kkaVa01ilNsgo4FtgbWA9c1Dx/I/At4ENt7UvSbBg+E9DWGYBtbXMc+5OWynw6nTyzKUkLa/NM6QeB24CvAq8F/hh4IHB4VV3R4n4kadEzCvdp5uHhWM45ZLdFLQceuM4g86l6y861pHFos1P6mKo6ACDJ+4AfA79SVXe0uA9JHepTB09qQ0+/ADCfzohRn6nb+sJq0tmhlbSQNjul9849qKr7knzPBCpJ0pKZTzVWs9Y53JH6XnnT7Rzdzy+vpKnSZqf0yUl+2jwOsGvzPEBV1aoW9yVJY7WtM2jjOGM8iWehe3qWcRqYT5doVMcBVuY9OWsdPEkah9Y6pVW1U1vbkiQtjpMzTR/zqdo0iV94TQu/sJAWbxz3KZUkjYkHmFL7+t556Or/fkf2u9DZ6z4bVd8TDuggEGkG2SmVppSdl37oy99hKUNt2zj76tBe9dWO/E/25f9ZkqaNnVJJmlJ9OoBe7P1hTzhg8xZnV+zMSoLFn82exLPKi92en4eaZnZKJXWqTx2nWbWUv0GXf69t7XuxZ363t6yk7pkXpNljp1Sd8UBR0rh4UCtJ0uSwUypNkEm5Fcn8bbYxUYSdjP6blL/RpMQpaTb4mSTZKV22JIcA/x3YCXhfVZ027n0u90OrjTOQW3cy7r/ua1vXhy13H8vd5lKG8I3D/Ovhhi32PpeeMd5xi/1bj2PYapdfHExbfbreppani/woSZpsdkqXIclOwF8DzwE2ApcmWV9VV3cb2WjjPljr0wFmnw9Mu+wEqH8m8XYJXfD/YbJMWn7U/fxfk9QlO6XLcxBwfVV9FyDJecBhgElXkjTLepUfu7r9S1f39+zTzLOLvWxjEv9GfVrX0VWaFqmqrmOYOEleBBxSVa9tnr8CeHpVHTdvuWOAY5qnjwe+s6KBjtcewI+7DqKHbJfRbJet2SajTUO7/GpVPbLrILrQYX6chvfNnGmpi/XoF+vRL7Ncj5E50jOly5MRZVv17qvqTODM8Yez8pJcVlVru46jb2yX0WyXrdkmo9kuE6+T/DhN75tpqYv16Bfr0S/WY2sPaGMjM2gjsO/Q832AmzuKRZKkvjA/SpKWzE7p8lwKrEmyX5IHAkcC6zuOSZKkrpkfJUlL5vDdZaiqzUmOAz7HYMr7s6rqqo7DWmlTOSy5BbbLaLbL1myT0WyXCdZhfpym98201MV69Iv16BfrMY8THUmSJEmSOuPwXUmSJElSZ+yUSpIkSZI6Y6dUS5bkjUkqyR5DZScluT7Jd5I8r8v4VlqSP01ybZJ/SfKJJA8bem1m2wUgySFN3a9PcmLX8XQlyb5JvpTkmiRXJTm+KX9EkouSXNf8fnjXsa60JDsl+WaSzzTPZ75NtHhJTklyU5Irmp/fHXpt4j5/Jz2/Jjm1yYVXJPl8kkcPvTZJ9ZiavJ7kxU3e+UWStfNem7S6TOQxRZKzktyS5NtDZROX68Z9LGOnVEuSZF/gOcD3h8r2ZzDD4hOAQ4C/SbJTNxF24iLgiVX1JOB/ASeB7dLU9a+B5wP7Ay9r2mQWbQZOqKp/CzwDOLZpixOBi6tqDXBx83zWHA9cM/TcNtFSnV5VT2l+/hEm8/N3SvLrn1bVk6rqKcBngLfBRNZjmvL6t4EXAl8eLpy0ukz4McU5DNp42CTmurEey9gp1VKdDryJLW+GfhhwXlXdU1XfA64HDuoiuC5U1eeranPz9BIG9+WDGW8XBnW9vqq+W1U/B85j0CYzp6o2VdU3msd3MOiE7c2gPc5tFjsXOLyTADuSZB/gBcD7hopnuk3Umkn8/J34/FpVPx16uhv312XS6jE1eb2qrqmq74x4adLqMrHHFFX1ZeDWecUTl+vGfSxjp1SLluRQ4Kaq+ta8l/YGfjD0fGNTNoteDVzYPJ71dpn1+o+UZDXwG8DXgL2qahMMPuyBPTsMrQt/weAg/BdDZbPeJlq645phlmcNDRubqM+facqvSd6e5AfAy2nOlDKB9RgyrXl90uoyafFuz0TnunEcy3ifUm0hyReAfzPipbcCbwGeO2q1EWVTda+hbbVLVX2qWeatDIY2fGhutRHLT1W7bMes138rSR4CfAx4Q1X9NBnVRLMhye8Bt1TV5UnWdRyOemw7eekM4FQGny2nAu9m0Ino3efPtOTX7eXDqnor8NYkJwHHASczgfVolpmIvL6YuoxabURZ53XZhkmLd2qN61jGTqm2UFW/M6o8yQHAfsC3mjffPsA3khzE4NuqfYcW3we4ecyhrqiF2mVOkqOA3wOeXfff/Hfq22U7Zr3+W0jySww+xD9UVR9vin+U5FFVtSnJo4BbuotwxT0TOLSZmOZBwKokf89st4lG2N7n75wk72VwHSP08PNnWvLrYv8ewD8AFzDolE5cPSYpry/hbzKsl3XZhkmLd3smMteN81jG4btalKq6sqr2rKrVVbWawYfDU6vqh8B64MgkuyTZD1gDfL3DcFdUkkOANwOHVtXPhl6a6XYBLgXWJNkvyQMZTKiwvuOYOpHBkeb7gWuq6s+HXloPHNU8PgpY6BvtqVNVJ1XVPs3nyZHAF6vq95nhNtHSNQdAc45gMKkLTNDn7zTl1yRrhp4eClzbPJ60esxCXp+0ukzbMcXE5bpxH8t4plQ7rKquSnI+cDWDYS7HVtV9HYe1kv4K2AW4qPmW+5Kq+qNZb5eq2pzkOOBzwE7AWVV1VcdhdeWZwCuAK5Nc0ZS9BTgNOD/JaxjMuPnibsLrFdtES/GuJE9hMIzvBuAPYXry0gTW47Qkj2dwnfiNwB/BRNZjavJ6kiOA9wCPBC5IckVVPW/S6jLJxxRJPgysA/ZIspHB6IFJzHVjPZbJ/SMSJEmSJElaWQ7flSRJkiR1xk6pJEmSJKkzdkolSZIkSZ1xoqMVsscee9Tq1au7DqM1d911F7vttlvXYfSO7TKa7bI122S0aWiXyy+//MdV9ciu45gUbeTHaXjfzJmWuliPfrEe/TLL9VgoR9opXSGrV6/msssu6zqM1mzYsIF169Z1HUbv2C6j2S5bs01Gm4Z2SXJj1zFMkjby4zS8b+ZMS12sR79Yj36Z5XoslCMdvitJkiRJ6oydUkmSJElSZxy+K02J1SdesMXzG057QUeRSJIkSYvnmVJJkiRJUmfslEqSJEmSOmOnVJIkSZLUGTulkiRJkqTO2CmVJEmSJHXGTqkkSZIkqTN2SiVJkiRJnem0U5rk+MWUSZI0a8yRkqRZ0fWZ0qNGlB29mBWTnJXkliTfHip7RJKLklzX/H740GsnJbk+yXeSPG+o/MAkVzav/WWSNOW7JPlIU/61JKuH1jmq2cd1SUbVQZKkHbWsHGl+lCRNmk46pUleluTTwH5J1g/9fAn4ySI3cw5wyLyyE4GLq2oNcHHznCT7A0cCT2jW+ZskOzXrnAEcA6xpfua2+Rrgtqp6LHA68M5mW48ATgaeDhwEnDyc3CVJ2hEt5MhzMD9KkibIzh3t938Cm4A9gHcPld8B/MtiNlBVXx7+drZxGLCueXwusAF4c1N+XlXdA3wvyfXAQUluAFZV1VcBknwAOBy4sFnnlGZbHwX+qvmW+HnARVV1a7PORQwS9YcXE7ckSduxQznS/ChJmjSddEqr6kbgRuDftbzpvapqU7OPTUn2bMr3Bi4ZWm5jU3Zv83h++dw6P2i2tTnJ7cAvD5ePWGcLSY5h8C0ze+21Fxs2bFh2xfrmzjvvnKr6tKXLdjnhgM1bPO/T38f3y9Zsk9Fsl7HlyKnOj9P0vpmWuliPfrEe/WI9ttbVmVIAkryQwbCfPYE0P1VVq9re1Yiy2kb5ctfZsrDqTOBMgLVr19a6deu2G+ik2LBhA9NUn7Z02S5Hn3jBFs9veHk3cYzi+2Vrtslotsv9VihHTkV+nKb3zbTUxXr0i/XoF+uxta4nOnoXcGhVPbSqVlXV7juYbH+U5FEAze9bmvKNwL5Dy+0D3NyU7zOifIt1kuwMPBS4dRvbkiSpTW3mSPOjJKm3uu6U/qiqrmlxe+u5f7bCo4BPDZUf2cwYuB+DCRu+3gxluiPJM5rrYV45b525bb0I+GJVFfA54LlJHt5M4PDcpkySpDa1mSPNj5Kk3up0+C5wWZKPAJ8E7pkrrKqPb2/FJB9mMGnDHkk2Mpjx7zTg/CSvAb4PvLjZ3lVJzgeuBjYDx1bVfc2mXsdgpsJdGUzgcGFT/n7gg82kD7cymJ2Qqro1yanApc1y/2VuUgdJklq0rBxpfpQkTZquO6WrgJ8x+DZ1TgHb7ZRW1csWeOnZCyz/duDtI8ovA544ovxumqQ94rWzgLO2F6MkSTtgWTnS/ChJmjSddkqr6lVd7l+SpL4yR0qSZkXXs++ezYiZ+arq1R2EI0lSb5gjJUmzouvhu58Zevwg4AicqU+SJDBHSpJmRNfDdz82/LyZnOELHYUjSVJvmCMlSbOi61vCzLcG+JWug5AkqYfMkZKkqdT1NaV3MLheJs3vHwJv7jImSZL6wBwpSZoVXQ/f3b3L/UuS1FfmSEnSrOh6oiOSHAr8VvN0Q1V9ZlvLS5I0K8yRkqRZ0Ok1pUlOA44Hrm5+jk/yji5jkiSpD8yRkqRZ0fWZ0t8FnlJVvwBIci7wTeCkTqOSJKl75khJ0kxovVOaZA3wDmB/BvdVA6CqHrPAKg8Dbm0eP7TteCRJ6gtzpCRJWxvHmdKzgZOB04GDgVcxmDlwlHcA30zypWaZ38JvgCVJ08scKUnSPOO4pnTXqroYSFXdWFWnAM8atWBVfRh4BvDx5uffVdV5Y4hJkqQ+MEdKkjTPODqldyd5AHBdkuOSHAHsOWrB5rWfVdX6qvpUs+7hY4hJkqQ+MEdKkjTPODqlbwAeDLweOBB4BXDUAsueXFW3zz2pqn9lMKxJkqRp9AbMkZIkbaH1a0qr6tLm4Z0MrpXZllGd4q5nBJYkaSzMkZIkba21M6VJHpPkrCT/NclDkrw3ybeT/I8kqxdY7bIkf57k15r1TwcubysmSZL6wBwpSdLC2hy+ew5wKYNvfy8BrgWeD3wWOGuBdf4T8HPgI8D5wP8Fjm0xJkmS+uAczJGSJI3U5jCg3avqDIAk/7Gq3t2Uvz/JcaNWqKq7gBNbjEGSpD4yR0qStIA2z5T+IsnjkjwNeHCStQBJHgvs1OJ+JEmaNOZISZIW0OaZ0jcBnwZ+ARwOnJTkycAq4A9a3I8kSZPGHClJ0gJa65Q2NwN//FDRPyfZA7itqu5raz+SJE0ac6QkSQtrdWr5JKuAR1bV/waoqh835U+qqn8ZWu49QC20nap6fZtxSZLUNXOkJEmjtXlLmJcwmE3wY0muaq6bmXPOvMUvYzCt/YOApwLXNT9PAfzGWJI0VcyRkiQtrM0zpW8BDqyqTUkOAj6Y5C1V9XEgwwtW1bkASY4GDq6qe5vnfwt8vsWYJEnqA3OkJEkLaLNTulNVbQKoqq8nORj4TJJ9WHgY0qOB3YFbm+cPacokSZom5khJkhbQZqf0jiS/NnStzKYk64BPAk9YYJ3TgG8m+VLz/LeBU1qMSZKkPjBHSpK0gDY7pa9j6yFIdyQ5BHjJqBWq6uwkFwJPb4pOrKofthiTJEl9YI6UJGkBbd4S5lsLlN8LfGgbq94DbGIwocPjkjyuqr7cVlySJHXNHClJ0sJavSUMQJIXAu8E9mTwrXCAqqpVI5Z9LXA8sA9wBfAM4KvAs9qOS5KkrpkjJUnaWmu3hBnyLuDQqnpoVa2qqt1HJdvG8cDTgBur6mDgN4D/M4aYJEnqA3OkJEnzjKNT+qOqumaRy95dVXcDJNmlqq4FHj+GmCRJ6gNzpCRJ87Q+fBe4LMlHGMwoeM9cYXMvtvk2JnlYs+xFSW4Dbt7RAJLcANzB4Cbjm6tqbZJHAB8BVgM3AC+pqtua5U8CXtMs//qq+lxTfiCDm5rvCvwjcHxVVZJdgA8ABwI/AV5aVTfsaNzSLFh94gVbPL/htBd0FInUiU5zpPlRktRH4+iUrgJ+Bjx3qKyArRJuVR3RPDylmfL+ocBnW4rj4Kr68dDzE4GLq+q0JCc2z9+cZH/gSAZT8j8a+EIzkcR9wBnAMcAlDJLuIcCFDBL0bVX12CRHMrg+6KUtxS0t2vwOnqTe60OOND9Kknql9U5pVb1qKcsn+U1gTTP1/SOBvYHvtR0XcBiwrnl8LrABeHNTfl5V3QN8L8n1wEHNt8mrquqrTZwfAA5nkHQP4/57xX0U+KskqaqFboAuSVJfc6T5UZLUqdY6pUneVFXvSvIeBt/6bqGqXj9inZOBtQyukTkb+CXg74Fn7mA4BXw+SQF/V1VnAntV1aYmlk1J9myW3ZvBN71zNjZl9zaP55fPrfODZlubk9wO/DIw/M2zJElAr3Kk+VGS1Dttnimdm7jhshGvLfQN6REMZhP8BkBV3Zxk9xZieWazrT0ZXIdz7TaWzYiy2kb5ttbZcsPJMQyGN7HXXnuxYcOGbQY9Se68886pqk9bVrpdTjhg84Kv9envM9wu82PuU5wryf+h0aa4XfqSI6cyP07T+2Za6mI9+sV69Iv12FprndKq+nTz+9z5ryX5swVW+3kzMUI1y+3WUiw3N79vSfIJ4CDgR0ke1XwL/CjglmbxjcC+Q6vvw2AiiY3N4/nlw+tsTLIzg+t8bh0Rx5nAmQBr166tdevWtVG9XtiwYQPTVJ+2rHS7HL2Na0pvePnKxbE9w+0yP+Y+xbmS/B8abVrbpS85clrz4zS9b6alLtajX6xHv1iPrY3jljCjvGSB8vOT/B3wsCR/AHwBeO+O7CjJbnPfJDcJ/LnAt4H1wFHNYkcBn2oerweOTLJLkv2ANcDXm6FMdyR5RpIAr5y3zty2XgR80etlJEnLtCI50vwoSeqrccy+O8pWw3maRPYR4NeBnzK4ZuZtVXXRDu5rL+ATg82zM/APVfXZJJcySPCvAb4PvBigqq5Kcj5wNbAZOLaZWRDgddw/5f2FzQ/A+4EPNpM+3MpgdkJJkpZjpXKk+VGS1EttTnT0iIVeYkTCbYYkfbKqDgR2tCM6vN3vAk8eUf4T4NkLrPN24O0jyi8Dnjii/G6apC1J0vb0IUeaHyVJfdXmmdLLWXgChHsXWOeSJE+rqktbjEOSpL4xR0qStIA2JzrabxmrHQz8YZIbgbsYJOuqqie1FZckSV0zR0qStLA2h+/+elVdm+Spo16vqm+MKH5+W/uXJKmvzJGSJC2szeG7/5nBPcfePVQ2POPes+avUFU3trh/SZL6yhwpSdICWrslTFUd0zw8Azisqg4GvgTcDryxrf1IkjRpzJGSJC1sHLeE+ZOqOj/JbwLPYfCt8BnA08ewL0ljtPrEC7Z4fsNpL+goEmlqmCMlSZqntTOlQ+buYfYC4G+r6lPAA0ctmOTVSdaMIQZJkvrIHClJ0jzjOFN6U5K/A34HeGeSXVi487sa+P0kv8pguvyvAF+pqivGEJckSV0zR0qSNM84OqUvAQ4B/qyq/jXJo4A/HrVgVb0NIMmuwB80y/0FsNMY4pLUGB6W65BcaUWZIyVJmqf1TmlV/Qz4+NDzTcCmUcsm+RPgmcBDgG8ymOzhK23HJGn87OhK22eOlCRpa+M4U7oULwQ2AxcA/wRcUlV3dxuSJEm9YI6UJM2EcUx0tGhV9VTg2cDXGcxCeGWSf+4yJkmS+sAcKUmaFZ2eKU3yRODfA78NrAV+gEOTJEkyR0qSZkbXw3ffCXwZ+Evg0qq6t+N4JC3S/HuYSmqdOVKSNBM67ZRWlbOhSJI0gjlSkjQruh6+uwZ4B7A/8KC58qp6TGdBSZLUA+ZISdKs6HSiI+Bs4AwGswseDHwA+GCnEUmS1A/mSEnSTOi6U7prVV0MpKpurKpTgGd1HJMkSX1gjpQkzYSuJzq6O8kDgOuSHAfcBOzZcUzSTHMCI6k3zJGSpJnQ9ZnSNwAPBl4PHAi8AnhllwFJktQTb8AcKUmaAV3Pvntp8/BO4FVJdgZeCnytu6gkSeqeOVKSNCs66ZQmWQUcC+wNrAcuap6/EfgW8KEu4pKm1bQPyZ1fvxtOW9k7aQzvf6X3reljjpQkzZquzpR+ELgN+CrwWuCPgQcCh1fVFR3FJGlMrrzpdo6e8o6x1CJzpLZr1JeNfikmaVJ11Sl9TFUdAJDkfcCPgV+pqjs6ikdSD3R9xlPqCXOkesvOsKRx6KpTeu/cg6q6L8n3TLaSJAHmyJkxqoN3ziG7dbJfO5aSutRVp/TJSX7aPA6wa/M8QFXVqo7iktQj47hW0+s/NQHMkUu00CUCffofn/Zr+/vMTrjUf510Sqtqpy72K0mTwGHMK6ePbW2O1LSaxM7hqC88+h6zNIk6vSWMpPEZx7fyffqmf7GxLCXmbS3bxkHI8PZXYoiepMXZkc5S3ztaffrcnjV9f29IfWKnVJJ6YLkHjl0OR+7jWUZJ7bFDK2ml2CmVJEkzYxI7WpMY8yTeCmxUO59wQAeBSDPITqkkLVGXB4htnFGFbZ/VbOPsq2dRNY0msXM4LRY7FLarv5HvDWnH2CmVpDGahA7sSsS4rY7u8GsnHLB5i7MrbXSe7SBLk8UOnjR77JSqMx4o9kPXyb/r/WtpJqGTvdRlJalvnCRJs8ZO6TIlOQT478BOwPuq6rRx73O5B1nT9iE2js7spHSQJ+VAe36cbVyTMyl171LXbdT1/hdrUuKcVF3kR0nSZLNTugxJdgL+GngOsBG4NMn6qrq628hGG8cB2Pwhdm1o61q5Nix3m+NoFy3dSt8uZjnLLWUSkC7f413/f/VpCLK2b9Lyo9QHfn5JdkqX6yDg+qr6LkCS84DDAJOuJGmW9So/7sjBflfr7sistX2q72JHyPQp5klcd7Eju0a9r9oeFdbVkOO27zM8Sl9H0M2ZhuHeqaquY5g4SV4EHFJVr22evwJ4elUdN2+5Y4BjmqePB76zooGO1x7Aj7sOoodsl9Fsl63ZJqNNQ7v8alU9susgutBhfpyG982caamL9egX69Evs1yPkTnSM6XLkxFlW/Xuq+pM4Mzxh7PyklxWVWu7jqNvbJfRbJet2Saj2S4Tr5P8OE3vm2mpi/XoF+vRL9Zjaw9oYyMzaCOw79DzfYCbO4pFkqS+MD9KkpbMTunyXAqsSbJfkgcCRwLrO45JkqSumR8lSUvm8N1lqKrNSY4DPsdgyvuzquqqjsNaaVM5LLkFtstotsvWbJPRbJcJ1mF+nKb3zbTUxXr0i/XoF+sxjxMdSZIkSZI64/BdSZIkSVJn7JRKkiRJkjpjp1RLluSNSSrJHkNlJyW5Psl3kjyvy/hWWpI/TXJtkn9J8okkDxt6bWbbBSDJIU3dr09yYtfxdCXJvkm+lOSaJFclOb4pf0SSi5Jc1/x+eNexrrQkOyX5ZpLPNM9nvk20eElOSXJTkiuan98dem3iPn8nPb8mObXJhVck+XySRw+9Nkn1mJq8nuTFTd75RZK1816btLpM5DFFkrOS3JLk20NlE5frxn0sY6dUS5JkX+A5wPeHyvZnMMPiE4BDgL9JslM3EXbiIuCJVfUk4H8BJ4Ht0tT1r4HnA/sDL2vaZBZtBk6oqn8LPAM4tmmLE4GLq2oNcHHzfNYcD1wz9Nw20VKdXlVPaX7+ESbz83dK8uufVtWTquopwGeAt8FE1mOa8vq3gRcCXx4unLS6TPgxxTkM2njYJOa6sR7L2CnVUp0OvIktb4Z+GHBeVd1TVd8DrgcO6iK4LlTV56tqc/P0Egb35YMZbxcGdb2+qr5bVT8HzmPQJjOnqjZV1Teax3cw6ITtzaA9zm0WOxc4vJMAO5JkH+AFwPuGime6TdSaSfz8nfj8WlU/HXq6G/fXZdLqMTV5vaquqarvjHhp0uoysccUVfVl4NZ5xROX68Z9LGOnVIuW5FDgpqr61ryX9gZ+MPR8Y1M2i14NXNg8nvV2mfX6j5RkNfAbwNeAvapqEww+7IE9OwytC3/B4CD8F0Nls94mWrrjmmGWZw0NG5uoz59pyq9J3p7kB8DLac6UMoH1GDKteX3S6jJp8W7PROe6cRzLeJ9SbSHJF4B/M+KltwJvAZ47arURZVN1r6FttUtVfapZ5q0MhjZ8aG61EctPVbtsx6zXfytJHgJ8DHhDVf00GdVEsyHJ7wG3VNXlSdZ1HI56bDt56QzgVAafLacC72bQiejd58+05Nft5cOqeivw1iQnAccBJzOB9WiWmYi8vpi6jFptRFnnddmGSYt3ao3rWMZOqbZQVb8zqjzJAcB+wLeaN98+wDeSHMTg26p9hxbfB7h5zKGuqIXaZU6So4DfA55d99/8d+rbZTtmvf5bSPJLDD7EP1RVH2+Kf5TkUVW1KcmjgFu6i3DFPRM4tJmY5kHAqiR/z2y3iUbY3ufvnCTvZXAdI/Tw82da8uti/x7APwAXMOiUTlw9JimvL+FvMqyXddmGSYt3eyYy143zWMbhu1qUqrqyqvasqtVVtZrBh8NTq+qHwHrgyCS7JNkPWAN8vcNwV1SSQ4A3A4dW1c+GXprpdgEuBdYk2S/JAxlMqLC+45g6kcGR5vuBa6rqz4deWg8c1Tw+CljoG+2pU1UnVdU+zefJkcAXq+r3meE20dI1B0BzjmAwqQtM0OfvNOXXJGuGnh4KXNs8nrR6zEJen7S6TNsxxcTlunEfy3imVDusqq5Kcj5wNYNhLsdW1X0dh7WS/grYBbio+Zb7kqr6o1lvl6ranOQ44HPATsBZVXVVx2F15ZnAK4Ark1zRlL0FOA04P8lrGMy4+eJuwusV20RL8a4kT2EwjO8G4A9hevLSBNbjtCSPZ3Cd+I3AH8FE1mNq8nqSI4D3AI8ELkhyRVU9b9LqMsnHFEk+DKwD9kiykcHogUnMdWM9lsn9IxIkSZIkSVpZDt+VJEmSJHXGTqkkSZIkqTN2SiVJkiRJnbFTKkmSJEnqjJ1SSZIkSVJn7JRKkiRJkjpjp1SSJEmS1Jn/H3Jq7tuCtRvFAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1080x288 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Draw Fig S5A with whole gene data\n",
    "\n",
    "Totdata = list(pos_siLin.keys())\n",
    "print(Totdata[15:30])\n",
    "print(type(pos_siLin))\n",
    "Totresdata = wholedata(Totdata, Trxdata, longest_isoform, Exondata, CDSdata, Startdata, Stopdata, pos_siLin, pos_siLuc)\n",
    "\n",
    "factor = 1000\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (15,4)\n",
    "\n",
    "plt.subplot(221)\n",
    "plt.hist(Totresdata[2], bins=range(-50,50))\n",
    "plt.ylim(0,120000)\n",
    "plt.ylabel('siLuc\\nRaw read count\\n')\n",
    "plt.grid()\n",
    "\n",
    "plt.subplot(222)\n",
    "plt.hist(Totresdata[3], bins=range(-50,20))\n",
    "plt.ylim(0,120000)\n",
    "plt.ylabel('siLin28a\\nRaw read count')\n",
    "plt.grid()\n",
    "\n",
    "plt.subplot(223)\n",
    "plt.hist(Totresdata[0], bins=range(-50,50))\n",
    "plt.ylim(0,120000)\n",
    "plt.ylabel('siLin28a\\nRaw read count')\n",
    "plt.grid()\n",
    "\n",
    "plt.subplot(224)\n",
    "plt.hist(Totresdata[1], bins=range(-50,20))\n",
    "plt.ylim(0,120000)\n",
    "plt.ylabel('siLin28a\\nRaw read count')\n",
    "plt.grid()\n",
    "\n",
    "\n",
    "#gap1 120\n",
    "#gap2 17343\n",
    "#print(Trx['ENSMUSG00000028792.14'])\n",
    "#print(Start['ENSMUSG00000028792.14'])\n",
    "#print(Stop['ENSMUSG00000028792.14'])\n",
    "#print(Lin['ENSMUSG00000028792.14'])\n",
    "plt.savefig('Figure S5A_wholegene.png', dpi=500)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSMUSG00000026980.15', 'ENSMUSG00000037169.15', 'ENSMUSG00000060377.4', 'ENSMUSG00000079435.9', 'ENSMUSG00000046591.10', 'ENSMUSG00000038412.8', 'ENSMUSG00000070047.14', 'ENSMUSG00000021025.8', 'ENSMUSG00000027248.13', 'ENSMUSG00000030062.7', 'ENSMUSG00000063316.13', 'ENSMUSG00000060461.5', 'ENSMUSG00000041959.14', 'ENSMUSG00000025967.16', 'ENSMUSG00000061518.11', 'ENSMUSG00000079941.1', 'ENSMUSG00000059070.16', 'ENSMUSG00000004748.5', 'ENSMUSG00000037470.14', 'ENSMUSG00000023944.14']\n"
     ]
    }
   ],
   "source": [
    "#Finding sample genes to check the code above. The result of this code is copy-and-pasted to a list named Sample\n",
    "\n",
    "counter=0\n",
    "sample=[]\n",
    "for gene in pos_siLin:\n",
    "    if counter==0:\n",
    "        counter+=1\n",
    "        continue\n",
    "    else:\n",
    "        sample.append(gene)\n",
    "        counter+=1\n",
    "        if counter>20:\n",
    "            break\n",
    "print(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n",
      "{'ENSMUST00000044043.2': [21501416, 21501418, 3, '-']}\n",
      "\n",
      "{'ENSMUST00000044043.2': [21501029, 21501031, 3, '-']}\n",
      "\n",
      "\n",
      "{'ENSMUSE00000645009.2': [60383138, 60383303, 166, '-'], 'ENSMUSE00000342189.2': [60375858, 60376229, 372, '-'], 'ENSMUSE00000250226.1': [60368759, 60368929, 171, '-'], 'ENSMUSE00000250220.1': [60365011, 60365175, 165, '-'], 'ENSMUSE00000250212.1': [60358468, 60358578, 111, '-'], 'ENSMUSE00000250204.1': [60357277, 60357417, 141, '-'], 'ENSMUSE00000250194.2': [60355369, 60355560, 192, '-'], 'ENSMUSE00000165085.1': [60354405, 60354562, 158, '-'], 'ENSMUSE00000165117.1': [60352922, 60353032, 111, '-'], 'ENSMUSE00000165093.1': [60352354, 60352442, 89, '-'], 'ENSMUSE00000165115.1': [60352056, 60352225, 170, '-'], 'ENSMUSE00000165091.1': [60349858, 60350057, 200, '-'], 'ENSMUSE00000165084.1': [60348353, 60348495, 143, '-'], 'ENSMUSE00000165071.1': [60340153, 60340234, 82, '-'], 'ENSMUSE00000165095.1': [60338295, 60338363, 69, '-'], 'ENSMUSE00000165110.1': [60334446, 60334564, 119, '-'], 'ENSMUSE00000165074.1': [60332925, 60332956, 32, '-'], 'ENSMUSE00000165059.1': [60332319, 60332494, 176, '-'], 'ENSMUSE00000165068.1': [60331077, 60331144, 68, '-'], 'ENSMUSE00000165061.1': [60330726, 60330808, 83, '-'], 'ENSMUSE00000165101.1': [60329976, 60330115, 140, '-'], 'ENSMUSE00000165062.1': [60328390, 60328495, 106, '-'], 'ENSMUSE00000165060.1': [60327787, 60327947, 161, '-'], 'ENSMUSE00000165070.1': [60323819, 60323948, 130, '-'], 'ENSMUSE00000165105.1': [60321698, 60321924, 227, '-'], 'ENSMUSE00000165083.1': [60318308, 60318473, 166, '-'], 'ENSMUSE00000165099.1': [60316089, 60316256, 168, '-'], 'ENSMUSE00000165064.1': [60314157, 60314273, 117, '-'], 'ENSMUSE00000165081.1': [60311609, 60311803, 195, '-'], 'ENSMUSE00000250051.1': [60308139, 60308365, 227, '-'], 'ENSMUSE00000250046.1': [60306330, 60306498, 169, '-'], 'ENSMUSE00000165107.1': [60303578, 60303727, 150, '-'], 'ENSMUSE00000165082.1': [60300955, 60301077, 123, '-'], 'ENSMUSE00000165100.1': [60299047, 60299214, 168, '-'], 'ENSMUSE00000692047.1': [60292103, 60293938, 1836, '-']}\n",
      "{'ENSMUST00000112533.7': [60383229, 60383231, 3, '-']}\n",
      "\n",
      "{'ENSMUST00000112533.7': [60293760, 60293762, 3, '-']}\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Scratch for debugging\n",
    "\n",
    "i = 'ENSMUSG00000079941.1'\n",
    "\n",
    "Start = Startdata[i]\n",
    "Stop = Stopdata[i]\n",
    "siLin_position = pos_siLin[i]\n",
    "siLuc_position = pos_siLuc[i]\n",
    "\n",
    "Exon=Exondata[i]\n",
    "CDS=CDSdata[i]\n",
    "Trx_name = longest_isoform[i]\n",
    "start_position = Start[Trx_name][0]\n",
    "stop_position = Stop[Trx_name][0] \n",
    "start_exon = getExon(Exon, start_position)\n",
    "stop_exon = getExon(Exon, stop_position)\n",
    "\n",
    "#UTR5 = CDS[start_exon][0] - Exon[start_exon][0]\n",
    "print(stop_exon)\n",
    "\n",
    "#UTR3 = Exon[stop_exon][1]- CDS[stop_exon][1]\n",
    "\n",
    "\n",
    "print(Start)\n",
    "print()\n",
    "print(Stop)\n",
    "print()\n",
    "#print(siLin_position)\n",
    "print()\n",
    "#print(siLuc_position)\n",
    "\n",
    "i = 'ENSMUSG00000026980.15'\n",
    "\n",
    "Start = Startdata[i]\n",
    "Stop = Stopdata[i]\n",
    "siLin_position = pos_siLin[i]\n",
    "siLuc_position = pos_siLuc[i]\n",
    "\n",
    "Exon=Exondata[i]\n",
    "CDS=CDSdata[i]\n",
    "Trx_name = longest_isoform[i]\n",
    "start_position = Start[Trx_name][0]\n",
    "stop_position = Stop[Trx_name][0] \n",
    "start_exon = getExon(Exon, start_position)\n",
    "stop_exon = getExon(Exon, stop_position)\n",
    "\n",
    "print(Exon)\n",
    "print(Start)\n",
    "print()\n",
    "print(Stop)\n",
    "print()\n",
    "#print(siLin_position)\n",
    "print()\n",
    "#print(siLuc_position)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trxdata\n",
      "{'ENSMUST00000162897.1': [3205901, 3216344, 10444, '-'], 'ENSMUST00000159265.1': [3206523, 3215632, 9110, '-'], 'ENSMUST00000070533.4': [3214482, 3671498, 457017, '-']}\n",
      "{'ENSMUST00000208660.1': [3999557, 4409241, 409685, '-'], 'ENSMUST00000208793.1': [4290846, 4409241, 118396, '-'], 'ENSMUST00000027032.5': [4344146, 4360314, 16169, '-']}\n",
      "{'ENSMUST00000027035.9': [4490931, 4496413, 5483, '-'], 'ENSMUST00000195555.1': [4491250, 4496757, 5508, '-'], 'ENSMUST00000192650.5': [4491390, 4497354, 5965, '-'], 'ENSMUST00000116652.7': [4491713, 4496363, 4651, '-'], 'ENSMUST00000192505.1': [4492457, 4493604, 1148, '-'], 'ENSMUST00000191647.1': [4492458, 4496330, 3873, '-'], 'ENSMUST00000191939.1': [4492465, 4493735, 1271, '-'], 'ENSMUST00000192913.1': [4492467, 4496396, 3930, '-']}\n",
      "\n",
      "longest_isoform\n",
      "ENSMUST00000162897.1\n",
      "ENSMUST00000208660.1\n",
      "ENSMUST00000027035.9\n",
      "\n",
      "Exondata\n",
      "{'ENSMUSE00000858910.1': [3213609, 3216344, 2736, '-'], 'ENSMUSE00000866652.1': [3205901, 3207317, 1417, '-']}\n",
      "{'ENSMUSE00001378580.1': [4409170, 4409241, 72, '-'], 'ENSMUSE00001403780.1': [4352202, 4352837, 636, '-'], 'ENSMUSE00001396015.1': [4351910, 4352081, 172, '-'], 'ENSMUSE00001380053.1': [4311270, 4311433, 164, '-'], 'ENSMUSE00001377871.1': [4292926, 4293012, 87, '-'], 'ENSMUSE00001379434.1': [4284766, 4284898, 133, '-'], 'ENSMUSE00001379919.1': [4267469, 4267620, 152, '-'], 'ENSMUSE00001380048.1': [4261527, 4261605, 79, '-'], 'ENSMUSE00001382043.1': [4245031, 4245106, 76, '-'], 'ENSMUSE00001379965.1': [4243543, 4243619, 77, '-'], 'ENSMUSE00001379150.1': [4243417, 4243448, 32, '-'], 'ENSMUSE00001378202.1': [4243133, 4243262, 130, '-'], 'ENSMUSE00001381486.1': [4231053, 4231144, 92, '-'], 'ENSMUSE00001382087.1': [4228443, 4228619, 177, '-'], 'ENSMUSE00001377002.1': [4226611, 4226823, 213, '-'], 'ENSMUSE00001378749.1': [4206660, 4206837, 178, '-'], 'ENSMUSE00001379718.1': [4197534, 4197641, 108, '-'], 'ENSMUSE00001381401.1': [4170205, 4170404, 200, '-'], 'ENSMUSE00001381514.1': [4163855, 4163941, 87, '-'], 'ENSMUSE00001376619.1': [4148612, 4148744, 133, '-'], 'ENSMUSE00001378642.1': [4147812, 4147963, 152, '-'], 'ENSMUSE00001379236.1': [4142612, 4142766, 155, '-'], 'ENSMUSE00001377224.1': [4120015, 4120073, 59, '-'], 'ENSMUSE00001381635.1': [4119668, 4119712, 45, '-'], 'ENSMUSE00001378903.1': [4092617, 4092780, 164, '-'], 'ENSMUSE00001380904.1': [4041888, 4042107, 220, '-'], 'ENSMUSE00001380986.1': [4024736, 4024890, 155, '-'], 'ENSMUSE00001378728.1': [4019070, 4019148, 79, '-'], 'ENSMUSE00001381361.1': [4007656, 4007737, 82, '-'], 'ENSMUSE00001378424.1': [3999557, 3999617, 61, '-']}\n",
      "{'ENSMUSE00000471383.2': [4496291, 4496413, 123, '-'], 'ENSMUSE00000231976.3': [4495136, 4495942, 807, '-'], 'ENSMUSE00000436360.3': [4493772, 4493863, 92, '-'], 'ENSMUSE00000154039.3': [4493100, 4493466, 367, '-'], 'ENSMUSE00000154037.5': [4490931, 4492668, 1738, '-']}\n",
      "\n",
      "CDSdata\n",
      "ENSMUSG00000025900.13\n",
      "{'ENSMUSE00001378580.1': [4409170, 4409187, 18, '-'], 'ENSMUSE00001403780.1': [4352202, 4352837, 636, '-'], 'ENSMUSE00001396015.1': [4351910, 4352081, 172, '-'], 'ENSMUSE00001380053.1': [4311270, 4311433, 164, '-'], 'ENSMUSE00001377871.1': [4292926, 4293012, 87, '-'], 'ENSMUSE00001379434.1': [4284766, 4284898, 133, '-'], 'ENSMUSE00001379919.1': [4267469, 4267620, 152, '-'], 'ENSMUSE00001380048.1': [4261527, 4261605, 79, '-'], 'ENSMUSE00001382043.1': [4245031, 4245106, 76, '-'], 'ENSMUSE00001379965.1': [4243543, 4243619, 77, '-'], 'ENSMUSE00001379150.1': [4243417, 4243448, 32, '-'], 'ENSMUSE00001378202.1': [4243133, 4243262, 130, '-'], 'ENSMUSE00001381486.1': [4231053, 4231144, 92, '-'], 'ENSMUSE00001382087.1': [4228443, 4228619, 177, '-'], 'ENSMUSE00001377002.1': [4226611, 4226823, 213, '-'], 'ENSMUSE00001378749.1': [4206660, 4206837, 178, '-'], 'ENSMUSE00001379718.1': [4197534, 4197641, 108, '-'], 'ENSMUSE00001381401.1': [4170205, 4170404, 200, '-'], 'ENSMUSE00001381514.1': [4163855, 4163941, 87, '-'], 'ENSMUSE00001376619.1': [4148612, 4148744, 133, '-'], 'ENSMUSE00001378642.1': [4147812, 4147963, 152, '-'], 'ENSMUSE00001379236.1': [4142612, 4142766, 155, '-'], 'ENSMUSE00001377224.1': [4120015, 4120073, 59, '-'], 'ENSMUSE00001381635.1': [4119668, 4119712, 45, '-'], 'ENSMUSE00001378903.1': [4092617, 4092780, 164, '-'], 'ENSMUSE00001380904.1': [4041888, 4042107, 220, '-'], 'ENSMUSE00001380986.1': [4024736, 4024890, 155, '-'], 'ENSMUSE00001378728.1': [4019070, 4019148, 79, '-'], 'ENSMUSE00001381361.1': [4007656, 4007737, 82, '-'], 'ENSMUSE00001378424.1': [3999560, 3999617, 58, '-']}\n",
      "ENSMUSG00000025902.13\n",
      "{'ENSMUSE00000154039.3': [4493100, 4493406, 307, '-'], 'ENSMUSE00000154037.5': [4491719, 4492668, 950, '-']}\n",
      "ENSMUSG00000033845.13\n",
      "{'ENSMUSE00000450421.7': [4785573, 4785677, 105, '-'], 'ENSMUSE00001229301.1': [4783951, 4784105, 155, '-'], 'ENSMUSE00001268906.1': [4782568, 4782733, 166, '-'], 'ENSMUSE00001298601.1': [4777525, 4777648, 124, '-'], 'ENSMUSE00000771965.1': [4774455, 4774516, 62, '-']}\n",
      "\n",
      "Startdata\n",
      "{'ENSMUST00000208660.1': [4409185, 4409187, 3, '-']}\n",
      "{'ENSMUST00000027035.9': [4493404, 4493406, 3, '-']}\n",
      "{'ENSMUST00000130201.7': [4785675, 4785677, 3, '-']}\n",
      "\n",
      "Stopdata\n",
      "{'ENSMUST00000208660.1': [3999557, 3999559, 3, '-']}\n",
      "{'ENSMUST00000027035.9': [4491716, 4491718, 3, '-']}\n",
      "{'ENSMUST00000130201.7': [4774452, 4774454, 3, '-']}\n"
     ]
    }
   ],
   "source": [
    "#Scratch for checking the code\n",
    "\n",
    "print('Trxdata')\n",
    "count=0\n",
    "for i in Trxdata:\n",
    "    count+=1\n",
    "    print(Trxdata[i])\n",
    "    if count>=3:\n",
    "        break\n",
    "\n",
    "print()\n",
    "print('longest_isoform')\n",
    "    \n",
    "count=0\n",
    "for i in longest_isoform:\n",
    "    count+=1\n",
    "    print(longest_isoform[i])\n",
    "    if count>=3:\n",
    "        break\n",
    "        \n",
    "print() \n",
    "print('Exondata')\n",
    "\n",
    "count=0\n",
    "for i in Exondata:\n",
    "    count+=1\n",
    "    print(Exondata[i])\n",
    "    if count>=3:\n",
    "        break\n",
    "\n",
    "print()\n",
    "print('CDSdata')\n",
    "        \n",
    "count=0\n",
    "for i in CDSdata:\n",
    "    count+=1\n",
    "    print(i)\n",
    "    print(CDSdata[i])\n",
    "    if count>=3:\n",
    "        break\n",
    "\n",
    "print()\n",
    "print('Startdata')\n",
    "\n",
    "count=0\n",
    "for i in Startdata:\n",
    "    count+=1\n",
    "    print(Startdata[i])\n",
    "    if count>=3:\n",
    "        break\n",
    "\n",
    "print()\n",
    "print('Stopdata')\n",
    "\n",
    "count=0\n",
    "for i in Stopdata:\n",
    "    count+=1\n",
    "    print(Stopdata[i])\n",
    "    if count>=3:\n",
    "        break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
