{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/qbio/nest/alpaca/Tutorial\n"
     ]
    }
   ],
   "source": [
    "%cd /qbio/nest/alpaca/Tutorial\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "\n",
    "\n",
    "#Generating dictoinary containing gene tags and their counts\n",
    "\n",
    "def makedata(addr):\n",
    "\n",
    "    with open(addr) as file_data:\n",
    "\n",
    "        Result = {}\n",
    "        for line in file_data:\n",
    "            if line.startswith('@'):\n",
    "                continue\n",
    "            else:\n",
    "                tmp1 = line.split('\\t')\n",
    "                if tmp1[1] == 4:\n",
    "                    continue\n",
    "                else:\n",
    "                    tmp2 = tmp1[2].split('|')\n",
    "                    tmp1[2] = tmp2[1]\n",
    "                    if (tmp1[2] in Result):\n",
    "                        Result[tmp1[2]] = Result[tmp1[2]] + 1\n",
    "                    else:\n",
    "                        Result[tmp1[2]] = 1\n",
    "        return Result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "Rseq_untreated = makedata('Align_RNA-seq-untreated/Aligned.out.sam')\n",
    "Rseq_siLuc = makedata('Align_RNA-seq-siLuc/Aligned.out.sam')\n",
    "Rseq_siLin28a = makedata('Align_RNA-seq-siLin28a/Aligned.out.sam')\n",
    "CLIP_Lin28a = makedata('Align_Lin28a-CLIP-35L33G/Aligned.out.sam')\n",
    "RiboProf_siLin28a = makedata('Align_RiboProf-siLin28a/Aligned.out.sam')\n",
    "RiboProf_siLuc = makedata('Align_RiboProf-siLuc/Aligned.out.sam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff = 250\n",
    "\n",
    "#Calculating Lin28A CLIP enrichment\n",
    "\n",
    "Xelements = {}\n",
    "\n",
    "for a in CLIP_Lin28a:\n",
    "    if a in Rseq_untreated:\n",
    "        if (CLIP_Lin28a[a] > cutoff and Rseq_untreated[a] > cutoff):\n",
    "            Xelements[a] = math.log(CLIP_Lin28a[a]/Rseq_untreated[a], 2)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "#Calculating Ribosome density change upon Lin28a KO\n",
    "\n",
    "tmp1 = {}\n",
    "tmp2 = {}\n",
    "Yelements = {}\n",
    "\n",
    "for a in RiboProf_siLin28a:\n",
    "    if a in Rseq_siLin28a:\n",
    "        if (RiboProf_siLin28a[a] > cutoff and Rseq_siLin28a[a] > cutoff):\n",
    "            tmp1[a] = RiboProf_siLin28a[a]/Rseq_siLin28a[a]\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "for a in RiboProf_siLuc:\n",
    "    if a in Rseq_siLuc:\n",
    "        if (RiboProf_siLuc[a] > cutoff and Rseq_siLuc[a] > cutoff):\n",
    "            tmp2[a] = RiboProf_siLuc[a]/Rseq_siLuc[a]\n",
    "    else:\n",
    "        continue\n",
    "        \n",
    "for a in tmp1:\n",
    "    if a in tmp2:\n",
    "        Yelements[a] = math.log(tmp1[a]/tmp2[a], 2)\n",
    "    else:\n",
    "        continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "size of Xelements: \n",
      "4890\n",
      "size of Yelements: \n",
      "8332\n",
      "size of CC: \n",
      "67332\n",
      "ENSMUSG00000106106.3\n",
      "ENSMUSG00000106106.3\n",
      "ENSMUST00000013773\n"
     ]
    }
   ],
   "source": [
    "#Generating dictoinary containing gene tags and their location\n",
    "\n",
    "with open('GO_Terms/uniprot-organism__mus+musculus_.tab') as file_data:\n",
    "\n",
    "    CC = {}\n",
    "    count = 0\n",
    "    for line in file_data:\n",
    "        if '(Mouse)' not in line:\n",
    "            continue\n",
    "        else:\n",
    "            count += 1\n",
    "            tmp1 = line.split('\\t')\n",
    "            tmp2 = tmp1[3].split('GO:')\n",
    "            tmp3 = tmp1[2].split(';')\n",
    "            Gene = []           \n",
    "            location = []\n",
    "            code = ''\n",
    "            \n",
    "            for i in range(len(tmp3)-1):\n",
    "                Gene.append(tmp3[i][:18])\n",
    "            \n",
    "            if len(Gene)==0:\n",
    "                continue\n",
    "            \n",
    "            for i in range(len(tmp2)-1):\n",
    "                location.append(tmp2[i+1][:7])\n",
    "                \n",
    "            #We only need information whether given gene product is located in nucleus(GO:0005634), integral membrane(GO:0016021), and cytoplasm(GO:0005737).\n",
    "            if '0005634' in location:\n",
    "                code += '1'\n",
    "            else:\n",
    "                code += '0'\n",
    "                \n",
    "            if '0016021' in location:\n",
    "                code += '1'\n",
    "            else:\n",
    "                code += '0'\n",
    "                \n",
    "            if '0005737' in location:\n",
    "                code += '1'\n",
    "            else:\n",
    "                code += '0'\n",
    "                \n",
    "            code = str(code)\n",
    "                \n",
    "            \n",
    "            for a in Gene:\n",
    "                CC[a]=code\n",
    "                \n",
    "#Generate lists of data to plot\n",
    "\n",
    "Nucleus = []\n",
    "IntegralMembrane = []\n",
    "Cytoplasm = []\n",
    "\n",
    "print(\"size of Xelements: \")\n",
    "print(len(Xelements))\n",
    "print(\"size of Yelements: \")\n",
    "print(len(Yelements))\n",
    "print(\"size of CC: \")\n",
    "print(len(CC))\n",
    "\n",
    "for a in Xelements:\n",
    "    print(a)\n",
    "    break\n",
    "    \n",
    "for a in Yelements:\n",
    "    print(a)\n",
    "    break\n",
    "    \n",
    "for a in CC:\n",
    "    print(a)\n",
    "    break\n",
    "\n",
    "for a in Xelements:\n",
    "    if (a in Yelements and a[:len(a)-2] in CC):\n",
    "        tmp = [Xelements[a], Yelements[a]]\n",
    "        if CC[a[:-2]][0]=='1':\n",
    "            Nucleus.append(tmp)\n",
    "        if CC[a[:-2]][1]=='1':\n",
    "            IntegralMembrane.append(tmp)\n",
    "        if CC[a[:-2]][2]=='1':\n",
    "            Cytoplasm.append(tmp)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "Data_Nucleus = np.array(Nucleus)      \n",
    "Data_IntegralMembrane = np.array(IntegralMembrane)\n",
    "Data_Cytoplasm = np.array(Cytoplasm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "not enough values to unpack (expected 2, got 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-049f218d83c4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#Plot figure S6.A\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mData_Nucleus\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscatter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'blue'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'Nucleus'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mData_IntegralMembrane\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscatter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'red'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'Integral membrane'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: not enough values to unpack (expected 2, got 0)"
     ]
    }
   ],
   "source": [
    "#Plot figure S6.A\n",
    "x, y = Data_Nucleus.T\n",
    "plt.scatter(x,y,s=4,c='blue', label='Nucleus')\n",
    "x, y = Data_IntegralMembrane.T\n",
    "plt.scatter(x,y,s=4,c='red', label='Integral membrane')\n",
    "x, y = Data_Cytoplasm.T\n",
    "plt.scatter(x,y,s=4,c='green', label='Cytoplasm')\n",
    "\n",
    "plt.title('Linkage to localization')\n",
    "plt.xlabel('LIN28A CLIP enrichment(log_2)')\n",
    "plt.ylabel('Ribosome density change upon Lin28a knockdown(log_2)')\n",
    "plt.legend()\n",
    "plt.grid()\n",
    "plt.savefig('Figure S6A.png', dpi=500)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
